[GFC] Relax GFC constraints to reflect support for row auto-flow.

yulunwu2 opened the pull request:
#### 56e58e729d54a8bb964e7dab7ca1aa4540ae8579
<pre>
[GFC] Relax GFC constraints to reflect support for row auto-flow.
<a href="https://bugs.webkit.org/show_bug.cgi?id=302902">https://bugs.webkit.org/show_bug.cgi?id=302902</a>
&lt;<a href="https://rdar.apple.com/165099744">rdar://165099744</a>&gt;

Reviewed by Sammy Gill.

This PR implements implicit grid growth when items with definite
row positions cannot fit in the available columns. This change allows
the GFC to correctly process items locked to a given row.

This PR updates the constraints to reflect support for items locked to
a given row and adds explicit validation that GFC requires at least
one explicit position in both row and column dimensions.

Combined changes:
* Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
(WebCore::Layout::GridLayout::trackSizingFunctions):
* Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
(WebCore::Layout::ImplicitGrid::insertDefiniteRowItem):
(WebCore::Layout::ImplicitGrid::findFirstAvailableColumnPosition const):
* Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
(WebCore::LayoutIntegration::gridLayoutAvoidanceReason):
(WebCore::LayoutIntegration::printReason):

Canonical link: <a href="https://commits.webkit.org/304069@main">https://commits.webkit.org/304069@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dba5cc5f1d1eb7da68128d7631ffa348625048b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134402 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6866 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45628 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141935 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86389 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136272 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7465 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6733 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102731 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69990 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/86aea5b4-f4d4-410d-9dbd-157757faccd7) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137349 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5209 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120458 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83522 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2edd9b99-7e9e-48ff-9d51-1b9da13e9ad4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5070 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2708 "Passed tests") | [⏳ 🛠 wpe-cairo-libwebrtc ](https://ews-build.webkit.org/#/builders/WPE-Cairo-LibWebRTC-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114280 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38600 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144626 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6546 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39184 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111137 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6622 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5498 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111407 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4900 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116736 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60328 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20763 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6598 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34934 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6423 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70169 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6658 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6534 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -107,23 +107,51 @@ void ImplicitGrid::insertDefiniteRowItem(const UnplacedGridItem& unplacedGridIte
+        return findFirstAvailableColumnPosition(normalizedRowStart, normalizedRowEnd, columnSpan, startSearchColumn);
+    };
+
+    auto growGridToFit = [&]() {
```
I think this is probably ok but could we try to make a couple of changes to better define this helper:

1.) The name is `growGridToFit` but I think it should probably mention what or why we are trying to fit. Right now it's a bit generic/vague and we probably don't want this to get used in more scenarios than we are intending
2.) Building on top of 1 I think this type of lambda might benefit from taking in arguments rather than capturing everything by reference. IMO that extra bit of information will help make more clear the intent of the function

Another bit that might help which you might want to consider is having the lambda `ASSERT(!findColumnPosition())` since we are using this to expand the grid when we can't fit the item

yulunwu2 replied:
Done, thanks for the review!

sammygill commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -107,23 +107,51 @@ void ImplicitGrid::insertDefiniteRowItem(const UnplacedGridItem& unplacedGridIte
+
+        // Verify the found position doesn't overlap with existing items
+        ASSERT(isCellRangeEmpty(*columnPosition, *columnPosition + columnSpan, normalizedRowStart, normalizedRowEnd),
+            "After grid growth, placed item overlaps with occupied cells");
```
Might be nice to add some debug logic that checks we didn't add any columns that are more than necessary as a sanity check. I think this might just be a matter of iterating over the last column and making sure there is at least one item in it after we perform the insertion below?

yulunwu2 replied:
Done, thanks for the review!

yulunwu2 committed:
[GFC] Relax GFC constraints to reflect support for row auto-flow.
https://bugs.webkit.org/show_bug.cgi?id=302902
<rdar://165099744>

Reviewed by Sammy Gill.

This PR implements implicit grid growth when items with definite
row positions cannot fit in the available columns. This change allows
the GFC to correctly process items locked to a given row.

This PR updates the constraints to reflect support for items locked to
a given row and adds explicit validation that GFC requires at least
one explicit position in both row and column dimensions.

Combined changes:
* Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
(WebCore::Layout::GridLayout::trackSizingFunctions):
* Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
(WebCore::Layout::ImplicitGrid::insertDefiniteRowItem):
(WebCore::Layout::ImplicitGrid::findFirstAvailableColumnPosition const):
* Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
(WebCore::LayoutIntegration::gridLayoutAvoidanceReason):
(WebCore::LayoutIntegration::printReason):

Canonical link: https://commits.webkit.org/304069@main
